src/data.jl

```diff
@@ -888,10 +888,90 @@ function h5convert!(out::Pointers, ::DataTypeODR, f::JLDFile, T::DataType, wsess
     nothing
 end
 
+# takes in x
+# returns [x, parentmodule(x), parentmodule(parentmodule(x)), ...]
+function _parent_module_recursive(x)
+    result = Any[]
+    tmp = x
+    while true
+        if !isempty(result) && tmp == result[end]
+            return result
+        else
+            push!(result, tmp)
+            tmp = parentmodule(tmp)
+        end
+    end
+end
+
+# returns the result of searching for the type in the specified module m
+function _resolve_type_singlemodule(::ReadRepresentation{T,DataTypeODR()},
+                                    m,
+                                    parts,
+                                    mypath,
+                                    hasparams::Bool,
+                                    params) where T
+    for part in parts[2:end]
+        sym = Symbol(part)
+        if !isdefined(m, sym)
+            return hasparams ? UnknownType(mypath, params) : UnknownType(mypath)
+        end
+        m = getfield(m, sym)
+    end
+    if !isa(m, DataType) && !isa(m, UnionAll)
+        return hasparams ? UnknownType(mypath, params) : UnknownType(mypath)
+    end
+    return m
+end
+
+_is_not_unknown_type(x::UnknownType) = false
+_is_not_unknown_type(x) = true
+
+# returns the results of searching for the type in all possible module
+function _resolve_type_allmodules(rr::ReadRepresentation{T,DataTypeODR()},
+                                  f::JLDFile,
+                                  ptr::Ptr,
+                                  header_offset::RelOffset,
+                                  mypath,
+                                  hasparams::Bool,
+                                  params) where T
+    parts = split(mypath, '.')
+    sym = Symbol(parts[1])
+
+    m_1 = filter(x->nameof(x) == sym, keys(Base.module_keys))
+    m_2 = Any[]
+    for x in m_1
+        append!(m_2, _parent_module_recursive(x))
```

Comment:
I'm not certain I fully understand why you can just start at `Base.moduleroot`? Meaning, why do you have to try a sub-module and then try its parent?

Also, aren't all the modules in `Base.module_keys` toplevel?

If these concerns are incorrect, some comment(s) illustrating why could be helpful.

Reply:
Good catch. Yeah everything in `Base.module_keys` is toplevel, so there isn't any need to do "submodule first, then parent."

Reply:
This is resolved now.

> Meaning, why do you have to try a sub-module and then try its parent?

Yeah we don't need to do that. In the updated PR, we just do the root modules.

> Also, aren't all the modules in `Base.module_keys` toplevel?

Correct. So in the updated PR, we don't try to take the "parent module" of the modules in `Base.module_keys`.